ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestCase.java

```diff
@@ -294,4 +329,244 @@ private static String standardizeTopology(final String topology) {
+          if (!outputRecordsFromTest.containsKey(record.topic.getName())) {
+            outputRecordsFromTest.put(record.topic.getName(), new ArrayList<>());
+          }
+          outputRecordsFromTest.get(record.topic.getName()).add(
+              FakeKafkaRecord.of(record, new ProducerRecord(
+                  record.topic.getName(),
+                  null,
+                  record.timestamp(),
+                  record.key(),
+                  record.topic.getSerializer(schemaRegistryClient).serialize(
+                      record.topic.getName(),
+                      record.value()
+                  )
+              )));
+        }
+    );
+    return outputRecordsFromTest;
+  }
+
+  private static Topic getTopic(
+      final KafkaTopicClient kafkaTopicClient,
+      final KsqlTopic ksqlTopic,
+      final SchemaRegistryClient schemaRegistryClient,
+      final boolean ignoreSrException) throws IOException, RestClientException {
+    final TopicDescription topicDescription = kafkaTopicClient
+        .describeTopic(ksqlTopic.getKafkaTopicName());
+    return new Topic(
+        ksqlTopic.getKafkaTopicName(),
+        getSchema(ksqlTopic, schemaRegistryClient, ignoreSrException),
+        getSerdeSupplierForKsqlTopic(ksqlTopic),
+        topicDescription.partitions().size(),
+        topicDescription.partitions().get(0).replicas().size()
+    );
+  }
+
+  private static SerdeSupplier getSerdeSupplierForKsqlTopic(final KsqlTopic ksqlTopic) {
+    Objects.requireNonNull(ksqlTopic);
+    if (ksqlTopic.getValueSerdeFactory() instanceof KsqlJsonSerdeFactory
+        || ksqlTopic.getValueSerdeFactory() instanceof KsqlDelimitedSerdeFactory) {
+      return new StringSerdeSupplier();
+    } else if (ksqlTopic.getValueSerdeFactory() instanceof KsqlAvroSerdeFactory) {
+      return new AvroSerdeSupplier();
+    }
+    throw new KsqlException("Unsupported topic serde: "
+        + ksqlTopic.getValueSerdeFactory().getClass().getSimpleName());
+  }
+
+  private static Optional<Schema> getSchema(
+      final KsqlTopic ksqlTopic,
+      final SchemaRegistryClient schemaRegistryClient,
+      final boolean ignoreSrException) throws IOException, RestClientException {
+    if (!(ksqlTopic.getValueSerdeFactory() instanceof KsqlAvroSerdeFactory)) {
+      return Optional.empty();
+    }
+    try {
+      return Optional.of(new Schema.Parser().parse(
+          schemaRegistryClient.getLatestSchemaMetadata(ksqlTopic.getKafkaTopicName()
+              + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX).getSchema()
+      ));
+    } catch (final IOException | RestClientException e) {
+      if (ignoreSrException) {
+        return Optional.empty();
+      }
+      throw e;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Windowed getWindowed(final Windowed windowed, final WindowType windowType) {
+    final String windowString = windowed.key().toString();
+    if (!windowString.contains("@") && !windowString.contains("/")) {
```

Comment:
What is this check for?

Reply:
This is because of discrepancy between the window key serdes. I will be investigating this more.